### DISCUSSION TITLE
api call failed, not waiting for ('onCharacteristicRead', 42)

### DISCUSSION THREAD

**Discussion by 9acca9** [hbldh/bleak#1662] [Q&A] [answered]
Hi.
Why im getting that response?
By the way this characteristic with notification.
this is the code:
```
import asyncio
import bleak

address = "40:22:D8:6C:C3:C2" # Replaced with actual bluetooth address

async def main(address):
    print("Connecting to device...")
    async with bleak.BleakClient(address) as client:
        print("Connected")
        servicios = client.services
        
        datos = await client.read_gatt_char(42)
        print(datos)


asyncio.run(main(address))
```

thanks.

Edit:
with other code i get this info:
1```
0-09 23:50:57.272  5981  6027 I python  : service UUID  00001801-0000-1000-8000-00805f9b34fb
10-09 23:50:57.272  5981  6027 I python  : service UUID  00001800-0000-1000-8000-00805f9b34fb
10-09 23:50:57.272  5981  6027 I python  : service UUID  4fafc201-1fb5-459e-8fcc-c5c9c331914b
10-09 23:50:57.272  5981  6027 I python  : Characteristic UUID  beb5483e-36e1-4688-b7f5-ea07361b26a8
10-09 23:50:57.272  5981  6027 I python  : Characteristic HANDLE  42
10-09 23:50:57.272  5981  6027 I python  : Characteristic HANDLE WITH HEX  0x2a
10-09 23:50:57.272  5981  6027 I python  : Characteristic DESCRIPTOR  [<bleak.backends.p4android.descriptor.BleakGATTDescriptorP4Android object at 0x7e5a16f290>, <bleak.backends.p4android.descriptor.BleakGATTDescriptorP4Android object at 0x7e5a16f2d0>]
```

**Comment by 9acca9** [comment]
So, maybe because a notify characteristic i have to use  BleakClient.start_notify
So, now is working but... what is the meaning of 

```
beb5483e-36e1-4688-b7f5-ea07361b26a8 (Handle: 42): Unknown: bytearray(b'b2')
10-10 00:51:00.420 15787 15823 I python  : beb5483e-36e1-4688-b7f5-ea07361b26a8 (Handle: 42): Unknown: bytearray(b'b2')
10-10 00:51:07.292 15787 15823 I python  : beb5483e-36e1-4688-b7f5-ea07361b26a8 (Handle: 42): Unknown: bytearray(b'b1')
```
My code send that values, that are strings: b1, b2.

What is telling me with Unknown, what is expected to be know? and why i get (b'b2') instead of 'b2'?

thank you very much!

**Comment by 9acca9** [comment]
Welll... i get it... lol. Sorry, yes, im a super beginner.

**Comment by 9acca9** [comment]
No, wait. Lol, i dont fix it as i want.
I change the function to this:
```
def callback(sender: bleak.BleakGATTCharacteristic, data: bytearray):
    print(f"{sender}: {data}")
    print(data)
    print(type(data))
    text = data.decode(encoding="utf-8")
    print(text)
    print(type(text)) # Just to know the type, lol
```

And well, yes, in that way i get my string clean.
But, there is no way to change:
`def callback(sender: bleak.BleakGATTCharacteristic, data: bytearray):`
for:
`def callback(sender: bleak.BleakGATTCharacteristic, data: str):`
??
because that dont work.

**Comment by MarkusPiotrowski** [comment, marked as answer]
Python is a dynamically typed language, that means in contrast to C# or Java or .... you are not defining variables statically.
The syntax you are using in your function definitions shows 'type hints' and these are indeed only hints for someone who reads the code or code linters
So it doesn't matter if you write `data: bytearray` or `data: str`, you can even omit it and just write `(sender, data)`.
Also, the variables which are received by the callback function have the type that they have. You cannot magically convert them by just writing another type hint to them.
So the data come in as bytestring, and if you want them to be strings, you must decode them (as you did).
